chips/nrf52/src/pwm.rs

```diff
@@ -0,0 +1,243 @@
+use kernel::common::cells::VolatileCell;
+use kernel::common::regs::{self, ReadOnly, ReadWrite, WriteOnly};
+use kernel::common::StaticRef;
+use kernel::hil;
+use kernel::ReturnCode;
+use nrf5x;
+
+#[repr(C)]
+struct PwmRegisters {
+    _reserved0: [u8; 4],
+    /// Stops PWM pulse generation on all channels at the end of current PWM period
+    tasks_stop: WriteOnly<u32, TASK::Register>,
+    /// Loads the first PWM value on all enabled channels
+    tasks_seqstart: [WriteOnly<u32, TASK::Register>; 2],
+    /// Steps by one value in the current sequence on all enabled channels if DECODER.MO
+    tasks_nextstep: WriteOnly<u32, TASK::Register>,
+    _reserved1: [u8; 240],
+    /// Response to STOP task, emitted when PWM pulses are no longer generated
+    events_stopped: ReadWrite<u32, EVENT::Register>,
+    /// First PWM period started on sequence 0
+    events_seqstarted: [ReadWrite<u32, EVENT::Register>; 2],
+    /// Emitted at end of every sequence 0, when last value from RAM has been
+    /// applied to the wave counter
+    events_seqend: [ReadWrite<u32, EVENT::Register>; 2],
+    /// Emitted at the end of each PWM period
+    events_pwmperiodend: ReadWrite<u32, EVENT::Register>,
+    /// Concatenated sequences have been played the amount of times defined in LOOP.CNT
+    events_loopsdone: ReadWrite<u32, EVENT::Register>,
+    _reserved2: [u8; 224],
+    /// Shortcut register
+    shorts: ReadWrite<u32, SHORTS::Register>,
+    _reserved3: [u8; 252],
+    /// Enable or disable interrupt
+    inten: ReadWrite<u32, INTEN::Register>,
+    /// Enable interrupt
+    intenset: ReadWrite<u32, INTEN::Register>,
+    /// Disable interrupt
+    intenclr: ReadWrite<u32, INTEN::Register>,
+    _reserved4: [u8; 500],
+    /// PWM module enable register
+    enable: ReadWrite<u32, ENABLE::Register>,
+    /// Selects operating mode of the wave counter
+    mode: ReadWrite<u32, MODE::Register>,
+    /// Value up to which the pulse generator counter counts
+    countertop: ReadWrite<u32, COUNTERTOP::Register>,
+    /// Configuration for PWM_CLK
+    prescaler: ReadWrite<u32, PRESCALER::Register>,
+    /// Configuration of the decoder
+    decoder: ReadWrite<u32, DECODER::Register>,
+    /// Amount of playback of a loop
+    loopreg: ReadWrite<u32, LOOP::Register>,
+    _reserved5: [u8; 8],
+    seq0: PwmSeqRegisters,
+    _reserved6: [u8; 16],
+    seq1: PwmSeqRegisters,
+    _reserved7: [u8; 16],
+    psel_out: [VolatileCell<nrf5x::pinmux::Pinmux>; 4],
+}
+
+#[repr(C)]
+struct PwmSeqRegisters {
+    seq_ptr: VolatileCell<*const u16>,
```

Comment:
Can anyone confirm that is is incorrect and should be `VolatileCell<*const ()>`?

Reply:
I don't think it's `*const ()`, I think it would either be `u8`  or `u16` (can't quite tell from the datasheet what the width of an element in the sequence would be... e.g. is the unit of `seq_cnt` a byte or a half-word?)